File: problems/medium/jump_game.rb
# @param {Integer[]} nums
# @return {Boolean}
def can_jump(nums)
  # state, it stands for could it jump to this postion
  f = []
  i = 0
  while(i<nums.length)
    f[i] = false
    i += 1
  end
  f[0] = true
  i = 1
  while(i<nums.length)
    j = 0
    while(j < i)
      if(f[j] == true && j + nums[j] >= i)
        f[i] = true
        break
      end
      j += 1
    end
    i += 1
  end
  f[nums.length - 1]
end

File: problems/medium/jump_game.go
/*
 * @Author: hongwei.sun
 * @Date: 2021-01-22 18:45:52
 * @LastEditors: your name
 * @LastEditTime: 2024-03-10 00:32:58
 * @Description: file content
 */
// https://leetcode.com/problems/jump-game/
/*
Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Determine if you are able to reach the last index.

Example 1:

Input: [2,3,1,1,4]
Output: true
Explanation: Jump 1 step from index 0 to 1, then 3 steps to the last index.
Example 2:

Input: [3,2,1,0,4]
Output: false
Explanation: You will always arrive at index 3 no matter what. Its maximum
             jump length is 0, which makes it impossible to reach the last index.
*/
func canJump(nums []int) bool {
	f := make([]bool, len(nums))
	f[0] = true
	for i := 1; i < len(nums); i++ {
		for j := 0; j < i; j++ {
			// can jump to i from j
			f[i] = f[j] && nums[j] >= (i-j)
			if f[i] == true {
				break
			}
		}
	}
	return f[len(nums)-1]
}